lib/ansible/galaxy/collection/__init__.py

```diff
@@ -1203,10 +1203,17 @@ def _build_files_manifest_walk(b_collection_path, namespace, name, ignore_patter
 
     manifest = _make_manifest()
 
+    def _discover_relative_base_directory(b_path: bytes, b_top_level_dir: bytes) -> bytes:
+        if b_path == b_top_level_dir:
+            return b''
+        common_prefix = os.path.commonpath((b_top_level_dir, b_path))
+        b_rel_base_dir = os.path.relpath(b_path, common_prefix)
+        return b_rel_base_dir.lstrip(os.path.sep.encode())
```

Comment:
This seems more complicated than it needs to be. I don't think you need to strip off the left most path separator in the last line.

Why not snip the right-most separator off at the beginning of `build_collection`? Then `_build_files_manifest_walk` won't need a change.

Otherwise, the third party caller should normalize the path before calling it, like we do internally.

Reply:
I'm just not sure anymore since sticking assertions and transformations into private functions that are going called a lot might end up been unnecessarily expensive.
Purely from the architecture perspective, I see CLI apps as layered where the outer layer deals with validation and transformation and the inner kernel assumes that the incoming data is already sanitized and can rely on it without repeating the same checks/transformations.

Reply:
That makes sense to me. I just wouldn't have guessed what the issue was from this implementation. If we port our internals to take/use pathlib objects, that aspect of the normalization would be handled. While I think it's not good to be using our unsupported Python API, our `build_collection` docstring suggests it just needs to be a directory with no caveat, and it's a minor enough change that I don't mind either way, so I'll leave it to you.

Reply:
@webknjaz @s-hertel sorry for the late reply, these days are a bit hectic, I would take any path you suggest, I think moving both variables (the prefix and the base dir vars) outside the loop are a nice to have change to avoid some unnecessary executions, and about the interfaces I think extending the current logic to avoid stripping string would be more than enough ( https://github.com/ansible/ansible/pull/81619/files#diff-b1b46df93a963b73e0c4a4c7a598a35d3135076a5ecf23f1c07323be0005a5fcL1209 ) I think the tests should cover both paths as they are all valid, although we always normalize paths before building the collection the function should do what its supposed to do.

Reply:
@ccamacho I'll just merge this, I suppose. Though, bear in mind that this is still a private API so there's no guarantees regarding the operation of these internals in the future.

If you want to contribute more improvements, I'd welcome a PR replacing the internal paths representation with pathlib objects in the dependency resolution code component.